### DISCUSSION TITLE
Config file

### DISCUSSION THREAD

**Discussion by joelfiddes** [ArcticSnow/TopoPyScale#84] [General]
The config is getting more complex and I have noticed one issue which can be confusing: we have started to develop optional parameters, which gives us three possibilities depending on how coded:

1. Parameter name must be declared + parameter value must be declared
2. Parameter name must be declared but parameter value is optional (can be blank)
3. Parameter name can be not declared (completely missing from config)

```python
if self.config.sampling.points.ID_col: 
    df_centroids['point_id'] = df_centroids[self.config.sampling.points.ID_col]
```
This parameter is declared as optional in the config but fits scenario 2

This example is scenario 3:
https://github.com/ArcticSnow/TopoPyScale/commit/2c8e37556e750fe08345dbe493bdec29f44ae741

**Comment by joelfiddes** [comment]
At some point would be good to look at something like https://models.slf.ch/p/inishell-ng/ as in our experience config errors are main source of non-running sims or unexpected results with SLF models (i think I mentioned this somewhere else).

**Comment by joelfiddes** [comment]
Dependency between parameters is another tricky thing that can be handled by a config generator.

**Comment by joelfiddes** [comment]
Some parameters have (or could have) default values - others not ie scenario 2 where default is None (or could be another default value)

**Comment by ArcticSnow** [comment]
1. As a start, what about writing in the output netcdf attribute all the config parameters, and commit hash and package version?
2. The package can include routines to check config incompatibilities. This is quite straight forward and can be implemented in a class called config that load config, check missing values, and incompatibilities.
3. Initshell sounds great, but this would be one more piece of software to sync with the core dev. We already struggle keeping the doc up to date...

**Comment by ArcticSnow** [comment]
I implemented a little code to grab git commit hash if available and TopoPyScale version used when exporting to netcdf.
This is bug free from commit: 39f70306d1ddaf2dc52b242d161b350d3065eab9
